### DISCUSSION TITLE
DataPortal methods: asynchronous or synchronous?

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#3619] [Questions]
Hello everyone,

I wanted to ask if it is advisable to program DataPortal methods on the server side asynchronously, or if it's a complication that isn't justified?

Thank you

**Comment by rockfordlhotka** [comment]
It depends on whether you are using any async API calls in your code. You can implement a data portal operation method that returns `void` or `Task` and the data portal will invoke them either way.

```c#
[Fetch]
private void Fetch(...)
{
  // you can use sync APIs here
}

[Fetch]
private async Task FetchAsync(...)
{
  // you can use the await keyword here
}
```

Either way, your code generally has the same complexity, it is just a question of whether you need to use the `await` keyword.
